### ISSUE TITLE
Torch compile settings

### ISSUE THREAD

**Issue by hal-314** [haifeng-jin/keras-benchmarks#2]
Hi there,

First of all, thanks for the benchmark! It's very useful to see this nice comparison between keras back-ends and native Pytorch.

I have two doubts:
- Why did you choose to use 'reduce-overhead' and not 'max-autotune' to compile native torch code? Pytorch docs recommends max-autotune for the best performance.
- Torch keras backend is generally slower than native torch. Sometimes, it's more than 2 times slower. Is keras using torch compile? If so, which mode is using? 

Finally, it could be useful to add how much time takes to compile / first model run for every model with different backends. So, you can know which backend you can use for quick prototyping and which one to use for long training jobs.

Thanks

**Comment by bhack** [comment]
It is unfair as you need to use `fullgraph=true` in `pytorch.compile` or you are comparing also eager sub nodes vs TF and JAX full compiled code.

**Comment by bhack** [comment]
> Finally, it could be useful to add how much time takes to compile / first model run for every model with different backends. So, you can know which backend you can use for quick prototyping and which one to use for long training jobs.

Yes also the warm start curation is very important and eventually dynamic recompilation timing (if activated).

**Comment by bhack** [comment]
Just to make an example with fast-SAM as you are benchmarking the Meta research vanilla version:
https://pytorch.org/blog/accelerating-generative-ai/
https://github.com/pytorch-labs/segment-anything-fast

As you can see in the turoial `torch.compile` sectionand its related repo over vanilla SAM (that you are using in this benchmark) you need at least verify that you are not going to benchmark eager fallback code or recompilations using `TORCH_LOGS=” graph_breaks,recompiles`. You can also enforce no graph break with `fullgraph=true` but checking about recompilation is also important for benchmarks to be sure to not include dynamic recompilations in the captured total time.

This is the minimal check you need to do not only for SAM but for all the models when you are going to activate the compilation to be sure you are not measuring recompilation or eager code.

`tf.function` (TF without user explicit `tf.py_function` or `tf.numpy_function`) + XLA or JAX are forcing you to write compile compliant code so you oould not have the same approach with a simple `torch.compile` on vanilla code without specific checks. What I meant is that if you find a JAX version of a model it was already written/modified to be JAX compiled by design and so on average it is not a fair test vs unmodified networks in the case you introduce in TF `tf.function ` + `tf.py_function` + `tf.numpy_function`, partial compilations, or just plain `torch.compile` without specific checks that could have implicit recompilation or silent eager graph breaks.

This is a specific issue about how to correctly inform your user/developer on what it is going to happen in the backend with your cross-backend shared `model.compile(jit_compile=keras_utils.use_jit())` interface especially if Keras 3 scope it is really to be true vendor neutral framework without a specific framework bias.

If you want to hide backends details to your users/developer you need to find the best way to be unbiased, as much as possible, across backends frameworks when using your API.

**Comment by bhack** [comment]
Just to make an example with native SAM you have here in the benchmark:

You have ignored an explicit warning about `torch.set_float32_matmul_precision('high')` for performance.

When you introduce this:
-  `fit`: 
`330.9897756576538 ms/step` instead of your reported `1,233.25 ms/step`
- `predict`:
`374.2371869087219 ms/step` instead of your reported `1,476.87 ms/step`


![immagine](https://github.com/haifeng-jin/keras-benchmarks/assets/1710528/a443c251-0efc-4152-bebd-b07ccf539611)

So currently, same hardware, just following the explicit warning the native pytorch impl is the fastest forward and backward implementation.

But you still have some extra candidate.

With `TORCH_LOGS=” graph_breaks,recompiles` you can see that you have some well know graph breaks caused by:
https://github.com/pytorch/pytorch/issues/120814

@JackCaoG , your colleague at Google, has an RFC/WIP at https://github.com/pytorch/pytorch/issues/117394

Finally you are not using cudagraph cause (these error were already explicit without extra flags when you published the benchmarks):
```python
skipping cudagraphs due to ['non-cuda device in graph']
skipping cudagraphs due to index put with accumulate. Found from : 
   File "/workspace/bench/keras-benchmarks/benchmark/sam/torch/fit.py", line 41, in resume_in_train_fn
    y_pred = model(input_image)
  File "/root/.venv/torch/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/root/.venv/torch/lib/python3.11/site-packages/segment_anything/modeling/image_encoder.py", line 112, in forward
    x = blk(x)
  File "/root/.venv/torch/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/root/.venv/torch/lib/python3.11/site-packages/segment_anything/modeling/image_encoder.py", line 174, in forward
    x = self.attn(x)
  File "/root/.venv/torch/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl
    return forward_call(*args, **kwargs)
  File "/root/.venv/torch/lib/python3.11/site-packages/segment_anything/modeling/image_encoder.py", line 234, in forward
    attn = add_decomposed_rel_pos(attn, q, self.rel_pos_h, self.rel_pos_w, (H, W), (H, W))
  File "/root/.venv/torch/lib/python3.11/site-packages/segment_anything/modeling/image_encoder.py", line 349, in add_decomposed_rel_pos
    Rh = get_rel_pos(q_h, k_h, rel_pos_h)
  File "/root/.venv/torch/lib/python3.11/site-packages/segment_anything/modeling/image_encoder.py", line 322, in get_rel_pos
    return rel_pos_resized[relative_coords.long()]
 ```

For `predict` you are skipping cudagraph for:
```python
skipping cudagraphs due to ['mutated inputs']
```

/cc @lezcano as he interacted with you already on https://github.com/haifeng-jin/keras-benchmarks/issues/1
@ezyang  @Chillee  if interested in this thread as this benchmark was claiming a quite important general pytorch performance gap also on social media.

**Comment by lezcano** [comment]
Also, I didn't follow up in the first issue, but if you are using cuda graphs, you need to run the model twice during warmp-up for the tracing to happen. This is not currently done in these benchmarks.

**Comment by bhack** [comment]
@lezcano They cannot use cudnngraph with the Meta research vanilla SAM reference impl at training time cause it has one of the fallback for (probably a data type case?):
https://github.com/pytorch/pytorch/pull/105439

Also it seems that in the vanilla SAM  on compiled training something in the compiled graph was assigned to CPU and this another root cause of why cudnngraph is then automatically disabled.

For inference instead probably with a new pytorch version or nightly we could better identify what is the specific line for the mutated input. This is the related @ezyang ticket:
https://github.com/pytorch/pytorch/issues/117281

I hope that a framework like Keras, at least, could really interact with upstream projects tickets on warnings usability, compilers limits, model performance best practices and compiler rough edges or blind spots.

It will be more useful than too quickly claims  general performance gaps.

**Comment by haifeng-jin** [comment]
Hi Everyone,

Thanks for all the comments!
I am trying to make the comparison as fair as possible.
So, if you find any unfair settings, please do let us know.

Here are my replies:

For the initial questions of this issue:
> Why did you choose to use 'reduce-overhead' and not 'max-autotune' to compile native torch code? Pytorch docs recommends max-autotune for the best performance.

I mainly followed the suggestions in #1. The `reduce-overhead` seems a more transparent since `max-autotune` seems unclear what was done under the hood. You are welcome try with other modes and report the results.

> Torch keras backend is generally slower than native torch. Sometimes, it's more than 2 times slower. Is keras using torch compile? If so, which mode is using?

Keras is not using `torch.compile()` by default. We have added the eager mode numbers for PyTorch native.
We still have a gap, but much smaller. I think it is mainly because of Keras is using the ops on a more fine grained level instead of calling the fused ops.

Some other concerns from people:

1. "You have ignored an explicit warning about `torch.set_float32_matmul_precision('high')` for performance."

We use `float32` for SAM on all frameworks.
Set it to `"high"` instead of `"highest"` is lowering the precision for one framework and leads to unfair comparison.

2. "you are not using cudagraph [for SAM]"

It is because of the model implementation from Meta Research.
We don't change the source code from Meta Research.
Changing the code contradicts with our purpose of measuring the "out-of-the-box" performance.

The implementation is not optimal, but it is representative for a common way to use PyTorch.

So, we benchmark it, and be explicit about it in the post.
It is referred to as "less manually-optimized model".

**Comment by Chillee** [comment]
> Set it to "high" instead of "highest" is lowering the precision for one framework and leads to unfair comparison.

In this case, the default in Jax/tensorflow is to turn on tensorfloat32 by default. So, in order to have a fair comparison, you need to enable tensorfloat32 in PyTorch.

**Comment by bhack** [comment]
@haifeng-jin So just starting from point 1) do you mean that you have changed the JAX default  to be fair? See:
https://github.com/google/jax/issues/7010

About 2) at least I've expected a big `*` in the blog post and in social media if the specific model had some specific issue with explicit warnings (we are not talking about hidden internals).

In any case I don't have the bandwidth to re-test all but just with my SAM random sampling from the list it would be enough to explain what you have done on 1) in JAX (or TF) as It Is enough to have pytorch native faster then the others versions in the table.

**Comment by haifeng-jin** [comment]
> Also, I didn't follow up in the first issue, but if you are using cuda graphs, you need to run the model twice during warmp-up for the tracing to happen. This is not currently done in these benchmarks.

Ah. Sorry for the oversight. The code is updated. Blog post update on the way.
We only used one batch for the warmup.

**Comment by haifeng-jin** [comment]
> > Set it to "high" instead of "highest" is lowering the precision for one framework and leads to unfair comparison.
> 
> In this case, the default in Jax/tensorflow is to turn on tensorfloat32 by default. So, in order to have a fair comparison, you need to enable tensorfloat32 in PyTorch.

@Chillee
Thanks for pointing this out!
I didn't know this.

Is it just `torch.backends.cuda.matmul.allow_tf32 = True`?
If it is that simple, I can measure it again and update the results.

Just find another caveat. I think Keras is explicitly specifying the dtypes to the backend it is using.
So it passes `float32` to TF/JAX. So I wonder if in that case it still uses tensorfloat32.
Is there a way to check if TF/JAX is actually using tensorfloat32?

**Comment by bhack** [comment]
See also https://github.com/google/jax/issues/10413#issuecomment-1495419411

**Comment by bhack** [comment]
> torch.backends.cuda.matmul.allow_tf32 = True?

Yes and you need to double check that the related warning is not appearing anymore.

**Comment by lezcano** [comment]
Was the tf32 point fixed in the end?

**Comment by haifeng-jin** [comment]
Hi @lezcano,

Thanks for following up!

Keras use `"float32"` explicitly when creating tensors. ([code link](https://github.com/keras-team/keras/blob/a05ac12efad3309462853605e67944550b9b2f0a/keras/src/backend/tensorflow/core.py#L106)) ([code link](https://github.com/keras-team/keras/blob/a05ac12efad3309462853605e67944550b9b2f0a/keras/src/backend/common/variables.py#L509)) ([code link](https://github.com/keras-team/keras/blob/a05ac12efad3309462853605e67944550b9b2f0a/keras/src/backend/config.py#L20))

So, to my understanding, Keras enforces TF and JAX to use `"float32"` instead of leaving it to the default value of the corresponding backend.

Therefore, the benchmarking code uses the same `"float32"` dtype for different backends. There is no fix required.

Everyone is welcome to check the actual dtype during runtime (if there is a way to do so) and post the method and results here!

**Comment by lezcano** [comment]
I don't think that is correct. TF32 is not a type per-se in keras or PyTorch. Both of them use it as a mode to perform fast float32 x float32 multiplication. As Horace mentioned, I'm pretty sure that JAX uses this by default. OTOH, in PyTorch is off by default, so you'd need to execute the PyTorch code with `torch.backends.cuda.matmul.allow_tf32 = True` to have a fair comparison.

**Comment by haifeng-jin** [comment]
I see. And I trust your expertise in PyTorch.

However, I am unable to verify if TF32 is actually enabled for TF/JAX.
I would prefer to just turn it off completely for all frameworks with `export NVIDIA_TF32_OVERRIDE=0` as suggested [here](https://deeprec.readthedocs.io/en/latest/NVIDIA-TF32.html).

What do you think?

**Comment by lezcano** [comment]
either that or enable it in both. I think a more reasonable comparison, given that you are benchmarking neural networks, would be to enable it in both, as that's what would be closer to reality (even better, networks would be executed on bfloat16 or amp, but yeah).

You can probably check that it exercises the tf32 cores by profiling the program with ncu and then looking at the cublas/cutlass kernels that it calls.

**Comment by lezcano** [comment]
To make sure they both use the same precision, you can manually enable the same flag in PyTorch (see above) and JAX
```python
jax.config.update("jax_default_matmul_precision", "highest")
```

**Comment by haifeng-jin** [comment]
@lezcano Thank you so much for teaching me all these! I have just turned it off.

**To everyone:**

Unfortunately, I do not have bandwidth to keep working on these benchmarks.
It would be more responsible to just remove it rather than leaving them as they are.

Therefore,
**I removed all the code for native PyTorch/HuggingFace** part
(and only kept the Keras with different backends),
and had a new snapshot and release of the repo.

I saw a lot concerns from people about the code in this repo.
**I thought these benchmarks would help people gain insights about different frameworks,
but it only caused more trouble instead.**
It was mainly because of my limited understanding of the frameworks involved.
My sincere apologies if these benchmarks have caused any trouble for you.
